### DISCUSSION TITLE
When logging in with cookie, how do you redirect upon success?

### DISCUSSION THREAD

**Discussion by caseydierking** [fastapi-users/fastapi-users#601] [Q&A]
Looked through the docs, but maybe I'm missing something. When using cookie based authentication, I setup a route to POST form data to the correct router. However, I can't seem to figure out how I would go about redirect the user to a different page upon both successful login and unsuccessful login.

Any help is appreciated.

**Comment by frankie567** [comment]
Hi @caseydierking!

I understand what you want to do; however, I want to point that FastAPI is mostly designed to create REST API, not dynamic websites like you could do with Flask or Django. This is why FastAPI Users is mostly designed to work as a REST API.

## Recommended approach

Now, to make a redirection upon successful login, what I recommend is to handle it on the **frontend side**. You probably have a frontend with a framework like React or Vue (it's also quite easy to do with vanilla JavaScrip too) in which you trigger the login request. You can then get the request response and make a browser redirection upon success (or show an error message if not).

## If you really want a server redirection

You can overload the `get_login_response` method of the `CookieAuthentication` class.

https://github.com/frankie567/fastapi-users/blob/3afc2037f1534207187b175e8a7f578643b6c399/fastapi_users/authentication/cookie.py#L91-L106

As you see, the main goal of this logic is to correctly set the session cookie. You can do something like this to add a redirection status code and a location header:

```py
class RedirectCookieAuthentication(CookieAuthentication):
    async def get_login_response(self, user: BaseUserDB, response: Response) -> Any:
        await super().get_login_response(user, response)
        response.status_code = 303
        response.headers["Location"] = "/success-route"
```

And something similar for `get_logout_response`.

**Reply by caseydierking** [reply]
Sorry for the late response, but this is what I ended up doing. I ended up calling it on the front end and redirecting there.

Thanks!

**Reply by bert2002** [reply]
@caseydierking how do you use the class? I tried using your code, but without any success.

**Reply by frankie567** [reply]
@bert2002 What have you tried so far? If you copy/paste this class into your code and use it as authentication backend, you should get the redirection on successful login.

**Reply by bert2002** [reply]
Hi @frankie567, I added the class and then used it as authentication backend:

Example:
```
cookie_authentication = RedirectCookieAuthentication(secret=SECRET, lifetime_seconds=86400, cookie_name="somethingcool")
auth_backends.append(cookie_authentication)
```
But I am getting an exception:
```
...
    raw_response = await run_endpoint_function(
  File "/home/user/.local/lib/python3.8/site-packages/fastapi/routing.py", line 159, in run_endpoint_function
    return await dependant.call(**values)
  File "/home/user/.local/lib/python3.8/site-packages/fastapi_users/router/auth.py", line 40, in login
    return await backend.get_login_response(user, response, user_manager)
TypeError: get_login_response() takes 3 positional arguments but 4 were given
```
Is it maybe because of a newer version of fastapi?

**Reply by frankie567** [reply]
Methods have changed slight a bit in newer versions of FastAPI Users. This should look like more like this now:

```py
class RedirectCookieAuthentication(CookieAuthentication):
    async def get_login_response(self, user: BaseUserDB, response: Response, user_manager: UserManager) -> Any:
        await super().get_login_response(user, response, user_manager)
        response.status_code = 303
        response.headers["Location"] = "/success-route"
```

**Reply by egelados** [reply]
Hi there,
congratulations on your great project!
Maybe you could also provide an example of how this could be done with javascript?

**Comment by IbraheemTuffaha** [comment]
I found an easy way to do this native in the backend with fastapi-users.
Suppose you have the following endpoint:
``` python
@app.get("/login")
def login_page(request: Request):
    # some logic
    return "success" # or whatever you want to return
```
and in `my_app.models.users` you have defined the `current_active_user` (check [docs](https://fastapi-users.github.io/fastapi-users/12.1/usage/current-user/)):
``` python
current_active_user = api_users.current_user(active=True)
```
So now if you add `current_active_user` as a dependency and set `optional=True` then user will be `None` if user is not logged in, otherwise it's logged in:
``` python
...
from my_app.models.users import User, current_active_user
...

@router.get("/login")
def login_page(
    request: Request,
    user: User | None = Depends(current_active_user),
):
    if user is not None:
        return RedirectResponse(url="/success-route")
    # Some logic
    return "success"
```
and
``` python
current_active_user = api_users.current_user(optional=True, active=True)
```
